app/scripts/controllers/permissions/index.js

```diff
@@ -188,33 +212,54 @@ export class PermissionsController {
    */
   async legacyExposeAccounts (origin, accounts) {
 
-    const permissions = {
-      eth_accounts: {},
+    // accounts are validated by finalizePermissionsRequest
+    if (typeof origin !== 'string' || !origin.length) {
+      throw new Error('Must provide non-empty string origin.')
     }
 
-    await this.finalizePermissionsRequest(permissions, accounts)
+    const existingAccounts = await this.getAccounts(origin)
+
+    if (existingAccounts.length > 0) {
+      throw new Error(
+        'May not call legacyExposeAccounts on origin with exposed accounts.'
+      )
+    }
+
+    const permissions = await this.finalizePermissionsRequest(
+      { eth_accounts: {} }, accounts
+    )
 
-    let error
     try {
+
       await new Promise((resolve, reject) => {
-        this.permissions.grantNewPermissions(origin, permissions, {}, (err) => (err ? resolve() : reject(err)))
+        this.permissions.grantNewPermissions(
+          origin, permissions, {}, _end
+        )
+
+        function _end (err) {
+          if (err) {
+            reject(err)
+          } else {
+            resolve()
+          }
+        }
       })
-    } catch (err) {
-      error = err
-    }
 
-    if (error) {
-      if (error.code === 4001) {
-        throw error
-      } else {
-        throw ethErrors.rpc.internal({
-          message: `Failed to add 'eth_accounts' to '${origin}'.`,
-          data: {
-            originalError: error,
-            accounts,
-          },
-        })
-      }
+      this.notifyDomain(origin, {
```

Comment:
Huh - so this was missing beforehand, and no `accountsChanged` event was fired upon a legacy connection? Well that's not good - good catch.

Reply:
It's amazing what writing tests will uncover.